backends/s3/s3.go

```diff
@@ -382,6 +420,61 @@ func New(ctx context.Context, opt Options) (*Backend, error) {
 	return b, nil
 }
 
+// NewReader satisfies ReaderStorage and provides a read streaming interface to
+// a blob located on an S3 server.
+func (b *Backend) NewReader(ctx context.Context, name string) (io.ReadCloser, error) {
+	r, err := b.doLoadReader(ctx, name, true, true)
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+// NewReader satisfies ReaderStorage and provides a write streaming interface to
+// a blob located on an S3 server.
+func (b *Backend) NewWriter(ctx context.Context, name string) (io.WriteCloser, error) {
+	pr, pw := io.Pipe()
+	wrap := &writerWrapper{backend: b, pw: pw, ctx: ctx, name: name}
+	wrap.wg.Add(1)
+	go func() {
+		defer wrap.wg.Done()
+		wrap.info, wrap.err = b.doStoreReader(ctx, name, pr, -1, true, true)
+	}()
+	return wrap, nil
+}
+
+// A writerWrapper implements io.WriteCloser and is returned by (*Backend).NewWriter.
+type writerWrapper struct {
+	// Basic data we need for the transfer.
+	ctx     context.Context
+	backend *Backend
+	name    string
+
+	// Writer-only logic variables.
+	pw *io.PipeWriter
+	wg sync.WaitGroup
```

Comment:
Instead of using a WaitGroup, perhaps use:

```go
done chan struct{}
```

with a `defer close(w.done)` in the goroutine and in `Close` do this to wait:

```go
switch {
case <-w.done:
case <-w.ctx.Done():
    return w.ctx.Err()
}
```

This is simpler than this use of WaitGroup, and it will return immediately if the context was closed.